### PR TITLE
Refinement of partial charge computation functionality

### PR DESCRIPTION
**Pull request by jacob-no-cop**
This PR constitutes a major refinement of the partial charge computation functionality added to biotite by PR #260 by accounting for the fact that identification of the hybridisation state is erroneous in certain edge cases, e. g. a deprotonated carboxyl group. To be more precise, the implementation in #260 (identification of the hybridisation state via the amount of binding partners) would erroneously assign the deprotonated oxygen atom in the carboxyl group a hybridisation state of sp2 because of one remaining binding partner although the hybridisation remains unchanged. In summary, the amount of binding partners of a considered atom is ambiguous with respect to its hybridisation state. Therefore, the new implementation in this PR performs the identification of the hybridisation state via an unambiguous characteristic, namely the bond type (single, double, aromatic, etc.). Nevertheless, the former implementation is not completely abandoned, but used as fallback implementation in case that the bond type `any`, i. e. an unspecified bond type occurs accompanied by a respective user warning about the potential faultiness  of the result. Moreover, a test verifying the new implementation has been added to the respective test file.

### CODE REVIEW

**Comment by padix-key** [comment]
I removed the compiled module, and added `*.pyd` files to the `.gitignore`.

File: src/biotite/structure/charges.py [hunk 1]
```diff
@@ -277,15 +415,64 @@ def partial_charges(atom_array, iteration_step_num=6, charges=None):
...
     amount_of_binding_partners = np.count_nonzero(bonds != -1, axis=1)
+    # The maximum of a given row of the `types` array must be determined
+    # as this value reveals the hybridisation state
+    try:
+        bond_types = np.amax(types, axis=1)
+        zero_indices_in_first_dim = np.unique(
+            np.nonzero(types == 0)[0]
+        )
+        bond_types[zero_indices_in_first_dim] = 0
+        # As nitrogen can be involved in aromatic systems having both
```
Comment (padix-key) [inline src/biotite/structure/charges.py:429]: I do not quite understand , why special handling for nitrogen is necessary. Shouldn't  `bond_types = np.amax(types, axis=1)` already give the correct hybridization?

File: tests/structure/test_charges.py [hunk 2]
```diff
@@ -356,4 +378,98 @@ def test_correct_output_ions():
     sodium_array.bonds = BondList(sodium_array.array_length())
     with pytest.warns(None) as record:
         partial_charges(sodium_array, iteration_step_num=1)
-    assert len(record) == 0
\ No newline at end of file
+    assert len(record) == 0
```
Comment (padix-key) [inline tests/structure/test_charges.py:381]: Please also check for the correctly calculated partial charge

File: tests/structure/test_charges.py [hunk 3]
```diff
@@ -356,4 +378,98 @@ def test_correct_output_ions():
...
+
+
+def test_correct_output_charged_aa():
+    """
+    Identification of hybridisation state is primarily performed via the
+    bond type; if the bond type is `any`, it is resorted to the amount
```
Comment (padix-key) [inline tests/structure/test_charges.py:387]: More precise: `BondType.ANY`

File: tests/structure/test_charges.py [hunk 4]
```diff
@@ -356,4 +378,98 @@ def test_correct_output_ions():
...
+        nitr_charge_without_btype, abs=5e-4
+    )
+    # Oxygen of the hydroxyl group in the carboxyl group has the index 2
+    oxyg_charge_with_btype = part_charges_with_btype[2]
+    oxyg_charge_without_btype = part_charges_without_btype[2]
+    assert oxyg_charge_with_btype < oxyg_charge_without_btype
+    # Assert that difference between the two values is significant
+    difference_oxyg_charges = abs(oxyg_charge_with_btype
+        - oxyg_charge_without_btype)
+    assert difference_oxyg_charges > 3e-2
```
Comment (padix-key) [inline tests/structure/test_charges.py:475]: As this function is quite long, please add some empty lines at fitting places for better readability.

Comment (jacob-no-cop) [reply tests/structure/test_charges.py:475]: Done.

File: tests/structure/test_charges.py (hunk 3 above)
Comment (jacob-no-cop) [reply tests/structure/test_charges.py:387]: Done.

File: tests/structure/test_charges.py (hunk 2 above)
Comment (jacob-no-cop) [reply tests/structure/test_charges.py:381]: Done.

File: src/biotite/structure/charges.py [hunk 5]
```diff
@@ -330,5 +517,6 @@ def partial_charges(atom_array, iteration_step_num=6, charges=None):
                     divisor) * damping
                 charges[i] += charge_transfer
                 charges[j] -= charge_transfer
+
     partial_charges = charges
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:521]: Done.

File: src/biotite/structure/charges.py (hunk 1 above)
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:429]: I added a comment explaining the need to proceed this way.

File: src/biotite/structure/charges.py [hunk 6]
```diff
@@ -12,69 +12,133 @@
...
-import warnings
+from biotite.structure import AtomArray
 
 
-# Creating dictionary to retrieve parameters for 
-# electronegativity computation from
-# First level of dictionary represents atom name
-# Second level represents hybridisation state
-
+# Creating dictionary to retrieve parameters for  electronegativity
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:21]: Done.

File: src/biotite/structure/charges.py [hunk 7]
```diff
@@ -277,15 +415,64 @@ def partial_charges(atom_array, iteration_step_num=6, charges=None):
...
+            np.nonzero(types == 0)[0]
+        )
+        bond_types[zero_indices_in_first_dim] = 0
+        # As nitrogen can be involved in aromatic systems having both
+        # only single bonds as well as one double bond and no special
+        # values for atoms involved in aromatic systems are used, it
+        # must be determined whether the considered atom possesses only
+        # single bonds or a double bond as well
+        nitrogen_bond_types = types[elements == "N"]
+        if 5 in nitrogen_bond_types:
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:435]: Done.

File: src/biotite/structure/charges.py [hunk 8]
```diff
@@ -277,15 +415,64 @@ def partial_charges(atom_array, iteration_step_num=6, charges=None):
...
+            nitrogen_indices = np.unique(
+                np.where(elements == "N")[0]
+            )
+            for i in nitrogen_indices:
+                if 5 in types[i]:
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:440]: Done.

File: src/biotite/structure/charges.py [hunk 9]
```diff
@@ -12,69 +12,133 @@
...
+# identification of the hybridisation state is either performed via the
+# bond type (primary way of identification since erroneous results are
+# excluded) or via the amount of binding partners (erroneous if a
+# binding partner is lost or added whilst the hybridisation remains
+# unchanged as it is the case with acids or bases, e. g. the hydroxyl
+# group in a carboxyl group: When the proton involved in the hydroxyl
+# group is donated, the amount of binding partners of the remaining
+# oxygen is one; this would erroneously lead to an identification of the
+# hybridisation state as sp2 although it is still sp3)
 EN_PARAMETERS = {
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:37]: Done.

File: src/biotite/structure/charges.py [hunk 10]
```diff
@@ -106,39 +173,96 @@ def _get_parameters(elements, amount_of_binding_partners):
...
     """
+
     parameters = np.zeros((elements.shape[0], 3))
+
     has_atom_key_error = False
     has_valence_key_error = False
+    each_btype_equal_zero = False
+    some_btype_equal_zero = False
+
+    if (all(btype == 0 for btype in bond_types)
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:184]: Done.

File: src/biotite/structure/charges.py [hunk 11]
```diff
@@ -106,39 +173,96 @@ def _get_parameters(elements, amount_of_binding_partners):
...
+        and
+        bond_types.shape[0] != 0):
+        each_btype_equal_zero = True
+    elif (any(btype == 0 for btype in bond_types)
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:188]: Done.

File: src/biotite/structure/charges.py [hunk 12]
```diff
@@ -106,39 +173,96 @@ def _get_parameters(elements, amount_of_binding_partners):
...
+        and
+        bond_types.shape[0] != 0):
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:190]: Done.

File: src/biotite/structure/charges.py [hunk 13]
```diff
@@ -106,39 +173,96 @@ def _get_parameters(elements, amount_of_binding_partners):
...
             else:
                 unparam_valence_names.append(element)
-                unparametrized_valences.append(
-                    amount_of_binding_partners[i]
-                )
+                if btype == "btype_equal_zero":
+                    unparametrized_valences.append(
+                        str(amount_of_binding_partners[i])
+                        +
+                        " " * 31
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:230]: Done.

File: src/biotite/structure/charges.py [hunk 14]
```diff
@@ -256,8 +382,20 @@ def partial_charges(atom_array, iteration_step_num=6, charges=None):
...
+        "The type of parameter `atom_array` must be an AtomArray."
+        )
+    if not isinstance(iteration_step_num, int):
+        raise ValueError(
+        "The type of parameter `iteration_step_num` must be an integer."
+        )
+    if not isinstance(charges, np.ndarray) and charges is not None:
+        raise ValueError(
+        "The type of parameter `charges` must be an NumPy ndarray."
+        )
```
[the remark this answers is not in the document]
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:397]: Done.

File: src/biotite/structure/charges.py [hunk 15]
```diff
@@ -138,6 +137,78 @@
...
+    -------
+    bond_types: ndarray, dtype=int
+        The inserted array in which the BondType.AROMATIC is replaced by
+        the a BondType revealing the bond order in case of nitrogen
+        atoms.
+    """
+
+    nitrogen_indices = np.where(elements == "N")[0]
+    for i in nitrogen_indices:
+         if BondType.AROMATIC in types[i]:
```
Comment (padix-key) [inline src/biotite/structure/charges.py:183]: Please combine with boolean mask from above for higher performance and rename variable accordingly:
```python
aromatic_nitrogen_indices = np.where((elements == "N") & (types == BondType.AROMATIC))[0]
```

File: src/biotite/structure/charges.py [hunk 16]
```diff
@@ -138,6 +137,78 @@
...
+            # `single`
+            # Therefore, the presence of a third bond type
+            # indicates a sp3 hybridisation, whereas the absence
+            # of a third bond type can be either due to sp2
+            # hybridisation or deprotonation
+            # In order to account for this ambiguity, the charge
+            # is considered in case that a third bond type is
+            # not present
+            try:
+                considered_row[-3]
```
Comment (padix-key) [inline src/biotite/structure/charges.py:200]: Do you check here, whether `consider_row` has length 3? In this case I think its clearer to use `if len(...) >= 3 ... else ...`

File: src/biotite/structure/charges.py [hunk 17]
```diff
@@ -138,6 +137,78 @@
...
+                bond_types[i] = BondType.SINGLE
+            except IndexError:
+                nitrogen_charge = charges[i]
+                if nitrogen_charge == -1:
```
Comment (padix-key) [inline src/biotite/structure/charges.py:204]: Are there actual cases of formally negatively charged nitrogen atoms in aromatic rings?

**Review by padix-key — CHANGES_REQUESTED** [review]

**Comment by padix-key** [comment]
I think there are still some minor issues in the special nitrogen handling. Could you resolve these? I think after that the code is ready to be merged.

File: src/biotite/structure/charges.py (hunk 15 above)
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:183]: Done.

File: src/biotite/structure/charges.py (hunk 16 above)
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:200]: You are right, I'll correct it.

File: src/biotite/structure/charges.py (hunk 17 above)
Comment (jacob-no-cop) [reply src/biotite/structure/charges.py:204]: This is indeed the case. For further reference, please see doi:10.1021/jo01066a039 ("Acidity Constants of Some Arylimidazoles and Their Cations". The Journal of Organic Chemistry. 26 (8): 2789–2791).

Comment (padix-key) [reply src/biotite/structure/charges.py]: OK

**Review by padix-key — APPROVED** [review]

**Comment by padix-key** [comment]
Thanks for introducing the changes. It looks good to me now.

**Pull request merged**